src/QsCompiler/CompilationManager/ProjectManager.cs

```diff
@@ -19,6 +19,31 @@ namespace Microsoft.Quantum.QsCompiler.CompilationBuilder
 {
     public delegate void SendTelemetryHandler(string eventName, Dictionary<string, string?> properties, Dictionary<string, int> measures);
 
+    /// <summary>
+    /// Holds compilation configuration -- currently, just whether this is a notebook.
+    /// Warning: properties of an instance may change as we get more information from the client.
+    /// </summary>
+    public class BuildConfiguration
+    {
+        /// <summary>
+        /// Indicates whether code is compiled for a notebook (Azure Notebook, Jupyter Notebook, etc.).
+        /// This means inferring an implicit namespace{} block (and banning explicit namespace{}
+        /// blocks), special treatment of use directives, and ignoring magic commands (e.g. %simulate)
+        /// </summary>
+        public bool IsNotebook { get; set; }
```

Comment:
The answer to this question is very disturbing and reflects poorly on my parents' job raising me.

Throughout the language server, there's a nasty problem: objects whose logic depends on whether we are in a notebook are instantiated before we actually know we are in a notebook. That's because we only get the `initialize` message with the `isNotebook` flag _after_ things like the EditorState and ProjectManager are instantiated. This is a hack to deal with that. Classes get instantiated with a handle to this BuildConfiguration, and then once we know we're in a notebook (by receiving the `initialize` message), `LanguageServer.Initialize()` sets `EditorState.IsNotebook = true`, which in turn sets `BuildConfiguration.IsNotebook = true`.

I had considered a bunch of spaghetti `setIsNotebook()` methods in every class to propagate that, but that was more disturbing in my opinion. I had considered a callback too as you mentioned, but getting F# to understand that was driving me nuts, and Sarah steered me away from that

Please let me know if you have a better ideas, because I'm not really satisfied with this implementation, but I'm not sure how to improve it